### PR TITLE
Fix Edge Cases in Password Callback Handling

### LINKED ISSUE #8441: APP may crash in load_pkcs12
the following code indicates that `len` may be >= `PEM_BUFSIZE`, if so, the `tpass` will not be ended with null-terminator, finally probably leads to crash when calling strlen in `PKCS5_PBKDF2_HMAC` ( <--pkcs12_gen_gost_mac_key <-- pkcs12_gen_mac <-- PKCS12_verify_mac <-- PKCS12_parse <-- load_pkcs12 )
https://github.com/openssl/openssl/blob/f0e4a860d0b350e10a1ee3898445cac85af8ea16/apps/apps.c#L398-L399
https://github.com/openssl/openssl/blob/f0e4a860d0b350e10a1ee3898445cac85af8ea16/crypto/evp/p5_crpt2.c#L30-L35
I think the description of `pem_password_cb` is not so clear, app may fill the whole `buf` regardless of null-terminator, app has not been informed that the `size` includes length of null-terminator.
or else if the `buf` implicitly requires a null-terminator when it is defined as type of `char *`, why do we still judge `len < PEM_BUFSIZE` but not write directly as follows:
`len = (len > (PEM_BUFSIZE - 1)) ? (PEM_BUFSIZE - 1) : len;`
`tpass[len] = 0;`
image

### PR DESCRIPTION
**Pull request by erbsland-dev**
Fixes #8441: This PR fixes edge cases in password callback handling.

Modify the password callback handling to reserve one byte in the buffer for a null terminator, ensuring compatibility with legacy behavior that puts a terminating null byte at the end.

Additionally, validate the length returned by the callback to ensure it does not exceed the given buffer size. If the returned length is too large, the process now stops gracefully with an appropriate error, enhancing robustness by preventing crashes from out-of-bounds access.

This PR also introduces a test suite for the password callback mechanism used when reading or writing encrypted and PEM or DER encoded keys via a BIO in OpenSSL. The test is designed to cover various edge cases, particularly focusing on scenarios where the password callback might return unexpected or malformed data from user code.

By simulating different callback behaviors, including negative returns, zero-length passwords, passwords that exactly fill the buffer and wrongly reported lengths. Also testing for the correct behaviour of binary passwords that contain a null byte in the middle.

- [X] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
I am unsure whether this should go into all branches or not.

**Review by t8m — APPROVED** [review]

**Comment by erbsland-dev** [comment]
> I am unsure whether this should go into all branches or not.

My goal was to address the specific issues in the password callback handling, but I don’t have the broader context to judge the impact or importance of this fix across different branches. I’ll leave it to the maintainers to decide the appropriate branches for inclusion, and I’m happy to make any changes as needed based on your recommendations.

File: test/bio_pw_callback_test.c
Comment (mattcaswell) [inline test/bio_pw_callback_test.c:31]: You treat encrypted_key_data as just a pointer to a buffer of bytes below (e.g. `char *`) - but you've declared it here as a `BUF_MEM *` which is definitely not correct (a `BUF_MEM` is a structured type)

File: test/bio_pw_callback_test.c
Comment (mattcaswell) [inline test/bio_pw_callback_test.c:81]: Probably you mean `OPENSSL_free()` here (assuming `encrypted_key_data` is a buffer of bytes...see my other comment)

File: test/bio_pw_callback_test.c
Comment (mattcaswell) [inline test/bio_pw_callback_test.c:172]: Probably you mean `OPENSSL_free`

File: test/bio_pw_callback_test.c
Comment (mattcaswell) [inline test/bio_pw_callback_test.c:383]: `OPENSSL_zalloc` would be better than these two lines

File: test/bio_pw_callback_test.c
Comment (mattcaswell) [inline test/bio_pw_callback_test.c:33]: Not really sure why this is a global, rather than a local in each function that needs it

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Review by t8m — APPROVED** [review]

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by erbsland-dev** [comment]
@t8m I'm not sure if the failure of "GitHub CI / self-hosted (freebsd-13.2)" is related to this pull request or not.

**Comment by mattcaswell** [comment]
>  I'm not sure if the failure of "GitHub CI / self-hosted (freebsd-13.2)" is related to this pull request or not.

No. An intermittent failure we see from time to time which we can't track down.

File: test/bio_pw_callback_test.c
Comment (shahsb) [inline test/bio_pw_callback_test.c:28]:
```suggestion
static const char cb_magic = 'p';
```

Please use `const` for all of these constants.

File: test/bio_pw_callback_test.c
Comment (shahsb) [inline test/bio_pw_callback_test.c:23]: Minor/nit: It would be nice to use `size_of()` to calculate size and try to avoid hardcoding's wherever possible.

**Review by shahsb — CHANGES_REQUESTED** [review]

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@beldmit please reconfirm

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by t8m** [comment]
Merged to all the active branches. Thank you for your contribution.

**Pull request closed without merging**
